core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -195,6 +195,16 @@ default ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connec
         return emptyList();
     }
 
+    /**
+     * Gets all of the columns on the specified table, or an empty list if the columns cannot be enumerated.
+     *
+     * @throws RuntimeException if table handle is no longer valid
+     */
+    default List<ColumnHandle> getColumns(ConnectorSession session, ConnectorTableHandle tableHandle)
```

Comment:
I have mixed feelings about this one. It is nice that we are explicit about data structure semantics by using `List` where we want expose an ordered collection. 

On the other hand I see also drawbacks. Major one is that we now do not have a single source of truth. 
If connectors sometimes return inconsistent information via `getColumns` and `getColumnHandles` it will result hard to track failures.
Also with `getColumnHandles` in place, being used to old patters, we can still write code which iterates over map, which would no longer be correct.

Minor thing is that naming does not reflect the semantics, but I assume we could work out better names.

Reply:
Agree.

> we can still write code which iterates over map, which would no longer be correct.

It will be 'correct' to exactly the same level as it was before with 'unforced guarantee' that 'Map' is backed by 'ImmutableMap'.

Part of this discussion was here: https://github.com/trinodb/trino/pull/6745#discussion_r566084730

> Minor thing is that naming does not reflect the semantics, but I assume we could work out better names.

Happy to apply your suggestions.